lib/python/picongpu/plugins/plot_mpl/phase_space_visualizer.py

```diff
@@ -21,55 +22,89 @@ class Visualizer(BaseVisualizer):
+            (the path before ``simOutput/``).
+            If None, the user is responsible for providing run_directories
+            later on via set_run_directories() before calling visualize().
         ax: matplotlib.axes
         """
-        super().__init__(run_directory, ax)
-
-        # for unit-conversion from SI (taken from picongpu readthedocs)
+        # TODO: remove this and get it from the readers metadata
+        # since it seems only correct for species='e' anyway.
         self.mu = 1.e6
         self.e_mc_r = 1. / (9.1e-31 * 2.9979e8)
-        self.cbar = None
+        self.colorbars = None
+
+        super().__init__(PhaseSpaceData, run_directory, ax)
 
-    def _create_data_reader(self, run_directory):
+    def set_run_directories(self, run_directories):
         """
-        Implementation of base class function.
+        Overridden from base class. Need to create colorbar instances
+        """
+        super().set_run_directories(run_directories)
+        self._init_colorbars(run_directories)
+
+    def _init_colorbars(self, run_directories):
+        self.colorbars = [None] * len(run_directories)
+
+    def _init_colors(self, run_directories):
+        """
+        Overridden from base class. Create colormaps instead
+        of colors since this is more useful for imshow plots.
+        """
+        self.colors = get_different_colormaps(len(run_directories))
+
+    def _remove_colorbar(self, idx):
+        """
+        Remove the colorbar for plot obj at idx.
         """
-        return PhaseSpaceData(run_directory)
+        self.colorbars[idx].remove()
+        self.colorbars[idx] = None
 
-    def _create_plt_obj(self):
+    def _remove_plt_obj(self, idx):
+        """
+        Overridden from base class.
+        Remove the colorbars before removing the plot object.
+        This order is necessary since otherwise matplotlib complains.
+        """
+        self._remove_colorbar(idx)
+        # clear the plt_obj and set it to None
+        super()._remove_plt_obj(idx)
+
+    def _create_plt_obj(self, idx):
         """
         Implementation of base class function.
         Turns 'self.plt_obj' into a matplotlib.image.AxesImage object.
         """
-        dat, meta = self.data
-        self.plt_obj = self.ax.imshow(
+        dat, meta = self.data[idx]
+        self.plt_obj[idx] = self.ax.imshow(
             np.abs(dat).T * meta.dV,
             extent=meta.extent * [self.mu, self.mu, self.e_mc_r, self.e_mc_r],
             interpolation='nearest',
             aspect='auto',
             origin='lower',
-            norm=LogNorm())
-        self.cbar = plt.colorbar(self.plt_obj, ax=self.ax)
-        self.cbar.set_label(
-            r'$Q / \mathrm{d}r \mathrm{d}p$ [$\mathrm{C s kg^{-1} m^{-2}}$] ')
+            norm=LogNorm(),
+            cmap=self.colors[idx])
+        # create the colorbar
+        self.colorbars[idx] = plt.colorbar(self.plt_obj[idx], ax=self.ax)
```

Comment:
if you set an alpha < 1.0, maybe also do [this work-around (see *Notes* section)](https://matplotlib.org/api/_as_gen/matplotlib.pyplot.colorbar.html#matplotlib.pyplot.colorbar) by default:
```python
self.colorbars[idx].solids.set_edgecolor("face")
```

Reply:
I think if you assign an explicit *colorbar axes* with `cax` your iteration layout problem might be addressable. Nevertheless, that would mean you have to manually place axes for colorbars.